mailbox/plugin/quota-mailing/src/main/java/org/apache/james/eventsourcing/CommandDispatcher.java

```diff
@@ -35,18 +34,22 @@
     public interface Command {
     }
 
+    public interface CommandHandler<C extends Command> {
+        Class<C> getCommandClass();
+
+        List<? extends Event> handle(C c);
+    }
+
     private final EventBus eventBus;
-    private final Map<Class, Function<Command, List<? extends Event>>> handlers;
+    private final Map<Class, CommandHandler> handlers;
 
     public CommandDispatcher(EventBus eventBus) {
         this.eventBus = eventBus;
         this.handlers = new ConcurrentHashMap<>();
     }
 
-    public <C extends Command, HandlerT extends Function<C, List<? extends Event>>>
-        CommandDispatcher register(Class<C> type, HandlerT commandHandler) {
-
-        handlers.put(type, (Function<Command, List<? extends Event>>) commandHandler);
+    public CommandDispatcher register(CommandHandler commandHandler) {
+        handlers.put(commandHandler.getCommandClass(), commandHandler);
```

Comment:
we have to prevent overriding of a previous handler configuration : a command must have a single handler and should probably never be redefined

Reply:
true! We need an assertion checking that before (and a test ensuring the same thing!)